### PR TITLE
Positive and negative look-ahead assertions

### PR DESCRIPTION
**Pull request by milseman**
Compiler support by using save points.

### CODE REVIEW

File: Sources/_StringProcessing/Compiler.swift [hunk 1]
```diff
@@ -92,11 +100,69 @@ class Compiler {
...
+      save(restoringAt: success)
+      save(restoringAt: intercept)
+      <sub-pattern>    // failure restores at intercept
+      clearSavePoint   // remove intercept
+      <if negative>:
+        clearSavePoint // remove success
+      fail             // positive->success, negative propagates
+    intercept:
+      <if positive>:
+        clearSavePoint // remove result
```
Comment (milseman) [inline Sources/_StringProcessing/Compiler.swift:138]:
```suggestion
        clearSavePoint // remove success
```

File: Sources/_StringProcessing/Compiler.swift [hunk 2]
```diff
@@ -101,6 +109,60 @@ class Compiler {
     }
   }
 
+  func compileLookaround(
```
Comment (rxwei) [inline Sources/_StringProcessing/Compiler.swift:112]:
```suggestion
  func emitLookaround(
```
Nit: How about starting with "emit" as existing subroutines called from `emit(_:)` are named "emit...".

File: Sources/_StringProcessing/Compiler.swift [hunk 3]
```diff
@@ -101,6 +109,60 @@ class Compiler {
...
+
+    case .lookahead, .negativeLookahead:
+      break
+
+    default: fatalError("unreachable")
+    }
+
+    let positive = kind == .lookahead
+    assert(positive || kind == .negativeLookahead)
+    _ = """
```
Comment (rxwei) [inline Sources/_StringProcessing/Compiler.swift:128]: Can we use a comment instead? What's the benefit of using a string literal?

File: Sources/_StringProcessing/Compiler.swift [hunk 4]
```diff
@@ -79,9 +79,17 @@ class Compiler {
...
       break
 
-    // FIXME: This can't be right...
     case .group(let g):
-      try emit(g.child)
+      switch g.kind.value {
+
+      case .lookahead, .negativeLookahead,
+          .lookbehind, .negativeLookbehind:
+        try compileLookaround(g.kind.value, g.child)
```
Comment (rxwei) [inline Sources/_StringProcessing/Compiler.swift:87]: We could use a `Lookaround` enum in the AST to group these four cases. Then `compileLookaround(_:_:)` would not need to handle unreachable cases.

**Review by rxwei — APPROVED** [review]

File: Sources/_StringProcessing/Compiler.swift (hunk 3 above)
Comment (milseman) [reply Sources/_StringProcessing/Compiler.swift:128]: Looks prettier in the editor (and sometimes better indentation than a multi-line comment), but I can use comments

File: Sources/_StringProcessing/Compiler.swift (hunk 4 above)
Comment (milseman) [reply Sources/_StringProcessing/Compiler.swift]: Added a `var lookaroundKind: (forwards: Bool, positive: Bool)?` to `Group`. This is probably a little more useful than an enum, which we'd want to immediately destructure or add these APIs to anyways. If we have more uses for more semantic info here, we can make the enum.

**Comment by milseman** [comment]
@swift-ci please test linux platform

**Pull request merged**